nav2_waypoint_follower/include/nav2_waypoint_follower/waypoint_follower.hpp

```diff
@@ -98,32 +108,85 @@ class WaypointFollower : public nav2_util::LifecycleNode
+   *
+   * @tparam T action_server
+   * @tparam U poses
+   * @tparam V feedback
+   * @tparam Z result
+   * @param action_server
+   * @param poses
+   * @param feedback
+   * @param result
+   */
+  template<typename T, typename U, typename V, typename Z>
+  void followWaypointsLogic(
+    const T & action_server,
+    const U & poses,
+    const V & feedback,
+    const Z & result);
+
   /**
    * @brief Action server callbacks
    */
-  void followWaypoints();
+  void followWaypointsCallback();
 
   /**
-   * @brief Action client result callback
-   * @param result Result of action server updated asynchronously
+   * @brief send robot through each of GPS
+   *        point , which are converted to map frame first then using a client to
+   *        `FollowWaypoints` action.
+   *
+   * @param waypoints, acquired from action client
    */
-  void resultCallback(const rclcpp_action::ClientGoalHandle<ClientT>::WrappedResult & result);
+  void followGPSWaypointsCallback();
+
+  /**
+ * @brief Action client result callback
+ * @param result Result of action server updated asynchronously
+ */
+  template<typename T>
+  void resultCallback(const T & result);
 
   /**
    * @brief Action client goal response callback
    * @param goal Response of action server updated asynchronously
    */
-  void goalResponseCallback(const rclcpp_action::ClientGoalHandle<ClientT>::SharedPtr & goal);
+  template<typename T>
+  void goalResponseCallback(const T & goal);
+
+/**
+ * @brief given some gps_waypoints, converts them to map frame using robot_localization's service `fromLL`.
+ *        Constructs a vector of stamped poses in map frame and returns them.
+ *
+ * @param gps_waypoints
+ * @param parent_node
+ * @param fromll_client
+ * @return std::vector<geometry_msgs::msg::PoseStamped>
+ */
+  static std::vector<geometry_msgs::msg::PoseStamped> convertGPSWaypointstoPosesinMap(
+    const
+    std::vector<sensor_msgs::msg::NavSatFix> & gps_waypoints,
+    const rclcpp_lifecycle::LifecycleNode::SharedPtr & parent_node,
+    nav2_util::ServiceClient<robot_localization::srv::FromLL> & fromll_client);
 
-  // Our action server
+  // Common vars used for both GPS and cartesian point following
+  rclcpp::Node::SharedPtr client_node_;
+  std::vector<int> failed_ids_;
+  int loop_rate_;
+  bool stop_on_failure_;
+
+  // Our action server for waypoint following
   std::unique_ptr<ActionServer> action_server_;
   ActionClient::SharedPtr nav_to_pose_client_;
-  rclcpp::Node::SharedPtr client_node_;
   std::shared_future<rclcpp_action::ClientGoalHandle<ClientT>::SharedPtr> future_goal_handle_;
-  bool stop_on_failure_;
   ActionStatus current_goal_status_;
-  int loop_rate_;
-  std::vector<int> failed_ids_;
+
+  // Our action server for GPS waypoint following
+  std::unique_ptr<ActionServerGPS> gps_action_server_;
+  nav2_util::ServiceClient<robot_localization::srv::FromLL> from_ll_to_map_client_;
```

Comment:
if you make this a unique pointer, then you don't need the bogus initialization in the class constructor to stop the "no default constructor" errors in L29.